IFC_Adapter/CRUD/Read.cs

```diff
@@ -0,0 +1,117 @@
+
+using BH.Engine.Adapters.IFC;
+using BH.oM.Adapter;
+using BH.oM.Adapters.IFC;
+using BH.oM.Base;
+using BH.oM.Data.Requests;
+using BH.oM.Geometry;
+using System.Collections.Generic;
+using System.Linq;
+using Xbim.Common.Geometry;
+using Xbim.Ifc2x3.Interfaces;
+using Xbim.ModelGeometry.Scene;
+
+namespace BH.Adapter.IFC
+{
+    public partial class IfcAdapter : BHoMAdapter
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        protected override IEnumerable<IBHoMObject> Read(IRequest request, ActionConfig actionConfig = null)
+        {
+            if (request == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("The request is null, pull failed.");
+                return new List<IBHoMObject>();
+            }
+
+            // Get the toolkit-specific PullConfig
+            IfcPullConfig config = actionConfig as IfcPullConfig;
+            if (config == null)
+            {
+                config = new IfcPullConfig();
+                BH.Engine.Reflection.Compute.RecordNote("Config has not been specified, default config is used.");
+            }
+
+            // Get the settings
+            IfcSettings settings = this.IFCSettings.DefaultIfNull();
+
+            // Get the discipline coming from the request/PullConfig
+            Discipline? requestDiscipline = request.Discipline(config.Discipline);
+            if (requestDiscipline == null)
+            {
+                BH.Engine.Reflection.Compute.RecordError("Conflicting disciplines have been detected.");
+                return new List<IBHoMObject>();
+            }
+
+            Discipline discipline = requestDiscipline.Value;
+
+            // If instructed to pull mesh representations, the below variable will be assigned
+            List<XbimShapeInstance> shapeInstances = null;
+
+            if (config.PullMeshes)
+            {
+                if (m_3DContext == null)
+                {
+                    m_3DContext = new Xbim3DModelContext(m_LoadedModel);
+                    m_3DContext.CreateContext();
+                }
+
+                shapeInstances = m_3DContext.ShapeInstances().ToList();
+            }
+
+            List<IBHoMObject> result = new List<IBHoMObject>();
+            foreach (var entity in m_LoadedModel.IIfcEntities(request))  
+            {
+                IIfcProduct element = entity as IIfcProduct;
+                if (element == null)
+                    continue;
+
+                IEnumerable<IBHoMObject> converted = element.IFromIfc(discipline, settings);
+
+                // Pull mesh representations if requested in PullConfig
+                if (shapeInstances != null)
+                {
+                    List<Mesh> meshes = shapeInstances.Where(x => x.IfcProductLabel == element.EntityLabel).SelectMany(x => x.Meshes(m_3DContext)).ToList();
```

Comment:
I can't see that `shapeInstances` is changing within this loop, so could the extraction of the meshes and creation of the `Representation` be done outside the `foreach` loop to be done once and reused rather than recalculated with every iteration?

Reply:
`shapeInstances` is assigned only once, `meshes` is different per each element so it needs to be done inside the `foreach` loop - I do not see anything that could be taken outside of it without harming the logic.